File: src/main/java/org/gecko/application/GeckoIOManager.java
package org.gecko.application;

import java.io.File;
import java.io.IOException;
import java.util.List;
import javafx.scene.Scene;
import javafx.scene.control.Alert;
import javafx.scene.control.ButtonType;
import javafx.stage.FileChooser;
import javafx.stage.Stage;
import javafx.util.Pair;
import lombok.Getter;
import lombok.Setter;
import org.gecko.io.FileTypes;
import org.gecko.io.ProjectFileParser;
import org.gecko.io.ProjectFileSerializer;
import org.gecko.model.GeckoModel;
import org.gecko.viewmodel.GeckoViewModel;
import org.gecko.viewmodel.PositionableViewModelElement;

public class GeckoIOManager {
    private static GeckoIOManager instance;
    @Getter
    @Setter
    private GeckoManager geckoManager;
    @Getter
    @Setter
    private Stage stage;

    @Getter
    @Setter
    private static File file;

    public static GeckoIOManager getInstance() {
        if (instance == null) {
            instance = new GeckoIOManager();
            file = new File("json/untitled_gecko_project.json");
        }
        return instance;
    }

    public void createNewProject() {
        File newFile = saveFileChooser(FileTypes.JSON);
        if (newFile != null) {
            file = newFile;
            Gecko newGecko = new Gecko();
            geckoManager.setGecko(newGecko);
            stage.setScene(new Scene(geckoManager.getGecko().getView().getMainPane()));
            saveGeckoProject(file);
        }
    }

    public void loadGeckoProject() {
        File fileToLoad = openFileChooser(FileTypes.JSON);
        if (fileToLoad != null) {
            ProjectFileParser projectFileParser = new ProjectFileParser();
            Pair<GeckoModel, GeckoViewModel> geckoPair = null;
            try {
                geckoPair = projectFileParser.parse(fileToLoad);
            } catch (IOException e) {
                Alert alert = new Alert(Alert.AlertType.ERROR,
                    "Corrupted file. Could not read project from " + fileToLoad.getPath() + ".", ButtonType.OK);
                alert.showAndWait();
            }

            if (geckoPair != null) {
                Gecko newGecko = new Gecko(geckoPair.getKey(), geckoPair.getValue());
                geckoManager.setGecko(newGecko);
                stage.setScene(new Scene(geckoManager.getGecko().getView().getMainPane()));
                file = fileToLoad;

                List<PositionableViewModelElement<?>> generatedViewModelElements =
                    projectFileParser.getGeneratedViewModelElements();
                if (generatedViewModelElements != null && !generatedViewModelElements.isEmpty()) {
                    Alert alert = new Alert(Alert.AlertType.CONFIRMATION,
                        "Loaded file contains elements with missing attributes. "
                            + "Do you want to keep those elements with default attributes?", ButtonType.YES,
                        ButtonType.NO);
                    alert.showAndWait();
                    if (alert.getResult().equals(ButtonType.NO)) {
                        for (PositionableViewModelElement<?> viewModelElement : generatedViewModelElements) {
                            geckoManager.getGecko().getViewModel().deleteViewModelElement(viewModelElement);
                        }
                    }
                }
            }
        }
    }

    public void importAutomatonFile(File file) {

    }

    public void saveGeckoProject(File file) {
        ProjectFileSerializer projectFileSerializer = new ProjectFileSerializer();
        try {
            projectFileSerializer.createFile(geckoManager.getGecko().getModel(), geckoManager.getGecko().getViewModel(),
                file);
        } catch (IOException e) {
            Alert alert = new Alert(Alert.AlertType.ERROR, "Designated file could not be created.", ButtonType.OK);
            alert.showAndWait();
        }
    }

    public void exportAutomatonFile(File file) {

    }

    public File openFileChooser(FileTypes fileType) {
        FileChooser fileChooser = getNewFileChooser(fileType);
        return fileChooser.showOpenDialog(stage);
    }

    public File saveFileChooser(FileTypes fileType) {
        FileChooser fileChooser = getNewFileChooser(fileType);
        return fileChooser.showSaveDialog(stage);
    }

    private FileChooser getNewFileChooser(FileTypes fileType) {
        FileChooser fileChooser = new FileChooser();
        fileChooser.getExtensionFilters()
            .addAll(new FileChooser.ExtensionFilter(fileType.getFileDescription(), fileType.getFileNameRegex()));
        return fileChooser;
    }
}

File: src/main/kotlin/org/gecko/application/GeckoIOManager.kt
package org.gecko.application

import javafx.event.EventHandler
import javafx.scene.control.*
import javafx.stage.FileChooser
import javafx.stage.Stage
import javafx.stage.WindowEvent
import org.gecko.exceptions.GeckoException
import org.gecko.exceptions.ModelException
import org.gecko.io.*
import org.gecko.view.ResourceHandler
import org.gecko.viewmodel.GeckoViewModel
import java.io.File
import java.io.IOException

/**
 * Represents a manager for the I/O functionalities of the Gecko Graphic Editor, following the singleton pattern.
 * Provides methods for creating, loading and saving project files or importing and exporting a project to another file
 * format. Uses the IO package for serialization and parsing.
 */
object GeckoIOManager {
    var geckoManager: GeckoManager? = null
    var file: File? = null
    var stage: Stage? = null
        set(value) {
            field = value
            field?.onCloseRequest = EventHandler { e: WindowEvent ->
                try {
                    launchSaveChangesAlert()
                } catch (ex: GeckoException) {
                    e.consume()
                }
            }
        }


    /**
     * Attempts to create a new project and makes the user choose a file to save it to.
     */
    fun createNewProject() {
        val newFile = getSaveFileChooser(FileTypes.JSON)
        if (newFile != null) {
            file = newFile
            val newGecko: Gecko
            try {
                newGecko = Gecko()
            } catch (e: ModelException) {
                geckoManager!!.gecko.viewModel.actionManager.showExceptionAlert(e.message)
                return
            }
            geckoManager!!.gecko = newGecko
            saveGeckoProject(file!!)
        }
    }

    /**
     * Attempts to load a project from a file chosen that was either previously chosen or asks the user to choose a
     * file.
     */
    fun loadGeckoProject() {
        val fileToLoad = getOpenFileChooser(FileTypes.JSON)!!
        val projectFileParser = ProjectFileParser()
        val gvm: GeckoViewModel =
            try {
                projectFileParser.parse(fileToLoad)
            } catch (e: IOException) {
                val alert = Alert(
                    Alert.AlertType.ERROR,
                    ResourceHandler.corrupted_file + fileToLoad.path + ".", ButtonType.OK
                )
                alert.showAndWait()
                return
            }

        val newGecko = Gecko(gvm)
        geckoManager!!.gecko = newGecko
        file = fileToLoad
    }

    /**
     * Imports an automaton from a file chosen by the user.
     *
     * @param file The file to import the automaton from.
     */
    fun importAutomatonFile(file: File) {
        val automatonFileParser = AutomatonFileParser()
        val gvm: GeckoViewModel?
        try {
            gvm = automatonFileParser.parse(file)
        } catch (e: IOException) {
            val message: String =
                ResourceHandler.could_not_read_file + String.format(
                    "%s.%s%s",
                    file.path,
                    System.lineSeparator(), e.message
                )
            val alert = Alert(Alert.AlertType.ERROR, message, ButtonType.OK)
            alert.showAndWait()
            return
        }
        val newGecko = Gecko(gvm)
        geckoManager!!.gecko = newGecko
        this.file = null
    }

    /**
     * Saves the current project to a file chosen by the user.
     *
     * @param file The file to save the project to.
     */
    fun saveGeckoProject(file: File) {
        val projectFileSerializer = ProjectFileSerializer(geckoManager!!.gecko.viewModel)
        try {
            projectFileSerializer.writeToFile(file)
        } catch (e: IOException) {
            val alert =
                Alert(
                    Alert.AlertType.ERROR, ResourceHandler.could_not_write_file,
                    ButtonType.OK
                )
            alert.showAndWait()
        }
    }

    /**
     * Exports the current automaton to a file chosen by the user.
     *
     * @param file The file to export the automaton to.
     */
    fun exportAutomatonFile(file: File?) {
        val fileSerializer: FileSerializer = AutomatonFileSerializer(geckoManager!!.gecko.viewModel)
        try {
            fileSerializer.writeToFile(file!!)
        } catch (e: IOException) {
            val alert =
                Alert(
                    Alert.AlertType.ERROR, ResourceHandler.could_not_write_file,
                    ButtonType.OK
                )
            alert.showAndWait()
        }
    }

    fun getOpenFileChooser(fileType: FileTypes): File? {
        val fileChooser = getNewFileChooser(fileType)
        return fileChooser.showOpenDialog(stage)
    }

    fun getSaveFileChooser(fileType: FileTypes): File? {
        val fileChooser = getNewFileChooser(fileType)
        var result = fileChooser.showSaveDialog(stage) ?: return null
        if (!result.name.endsWith("." + fileType.fileExtension)) {
            result = File(result.path + "." + fileType.fileExtension)
        }
        return result
    }

    fun getNewFileChooser(fileType: FileTypes): FileChooser {
        val fileChooser = FileChooser()
        fileChooser.extensionFilters
            .addAll(FileChooser.ExtensionFilter(fileType.fileDescription, fileType.fileNameGlob))
        return fileChooser
    }

    @Throws(GeckoException::class)
    fun launchSaveChangesAlert() {
        val saveChangesAlert =
            Alert(
                Alert.AlertType.NONE, ResourceHandler.save_changes_prompt, ButtonType.YES,
                ButtonType.NO
            )
        saveChangesAlert.title = ResourceHandler.confirm_exit
        saveChangesAlert.showAndWait()

        if (saveChangesAlert.result == ButtonType.NO) {
            return
        }

        if (file == null) {
            val fileToSaveTo = getSaveFileChooser(FileTypes.JSON) ?: throw GeckoException("No file chosen.")
            file = fileToSaveTo
        }
        saveGeckoProject(file!!)
    }
}